 design/README.MD

```diff
@@ -0,0 +1,456 @@
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Smithy-Dafny modes
+
+### Library
+
+`smithy-dafny` supports `@localService` development.
+This is where the implementation is locally executing code.
+Rather than a remote service like most AWS Service Smithy models.
+
+In this mode `smithy-dafny` will produce Dafny types and interfaces
+for a Smithy model in Dafny.
+This creates an `abstract module` in Dafny you can refine
+with a concrete implementation.
+
+`smithy-dafny` also generates polymorph code
+that will expose native types and interfaces
+for your Smithy model.
+This creates a shim with native interfaces.
+These native interfaces are connected
+to the Dafny interface
+and takes care of type conversion back and forth.
+
+```mermaid
+---
+title: Smithy Dafny Library
+---
+flowchart RL
+subgraph Library[Native Library]
+    direction RL
+    Interface[Public Interface] --> ToDafny[Convert input \nto Dafny]
+    ToDafny --> DafnyImplementation[Dafny implementation]
+    DafnyImplementation --> ToNative[Convert output \nto native]
+    ToNative --> Interface
+end
+
+UserCode --> Interface
+Interface --> UserCode
+```
+
+The goal is for native developers
+to call Dafny libraries with idiomatic types.
+
+### Wrapped AWS Services
+
+`smithy-dafny` also supports wrapping existing AWS Services
+with existing implementations that have a native implementation.
+
+The problem we are trying to solve here is:
+I am writing code in Dafny
+and I want to call an AWS service,
+how do I do that?
+
+In this mode `smithy-dafny` will produce Dafny types and interfaces
+so that Dafny code can call the native implementation.
+
+`smithy-dafny` also generates polymorph code
+so that these Dafny types and interfaces
+are connected to the native interface.
+Along with all the required type conversions.
+
+Again there is an `abstract module` that you refine
+with a `extern` that returns a native SDK instance
+wrapped with the Dafny polymorph shim.
+Ideally these `extern` would be generated
+but the constructors for the SDKs
+are not described in Smithy.
+
+In this diagram we zoom in on
+only the Dafny implementation.
+
+```mermaid
+---
+title: Wrapped AWS service
+---
+flowchart RL
+subgraph Library[Dafny Library]
+    direction RL
+    DafnyImplementation[Dafny implementation]
+end
+
+subgraph WrappedSDK
+    direction RL
+    SdkInterface[Generated Dafny SDK Interface] --> ToNative[Convert input \nto native]
+    ToNative --> NativeSDK[Native SDK implementation]
+    NativeSDK --> ToDafny[Convert output \nto Dafny]
+    ToDafny --> SdkInterface
+end
+
+DafnyImplementation --> SdkInterface
+SdkInterface --> DafnyImplementation
+
+```
+
+The goal is for Dafny developers
+to call native AWS SDKs.
+
+### Wrapped Library
+
+`smithy-dafny` can wrap an existing `smithy-dafny` library.
+This is useful for integration testing.
+
+When you consider the first two use cases
```

Comment:
use cases == modes? maybe this could say e.g. "when you consider `smithy-dafny` use cases like local services and wrapped AWS services, it's clear that..." and / or maybe the `Smithy-Dafny modes` section could be renamed to `Smithy-Dafny use cases`?